indra/llconvexdecomposition/llconvexdecompositionvhacd.cpp

```diff
@@ -0,0 +1,492 @@
+{
+    return LLSimpleton::getInstance();
+}
+
+LLCDResult LLConvexDecompositionVHACD::initSystem()
+{
+    createInstance();
+    return LLCD_OK;
+}
+
+LLCDResult LLConvexDecompositionVHACD::initThread()
+{
+    return LLCD_OK;
+}
+
+LLCDResult LLConvexDecompositionVHACD::quitThread()
+{
+    return LLCD_OK;
+}
+
+LLCDResult LLConvexDecompositionVHACD::quitSystem()
+{
+    deleteSingleton();
+    return LLCD_OK;
+}
+
+LLConvexDecompositionVHACD::LLConvexDecompositionVHACD()
+{
+    //Create our vhacd instance and setup default parameters
+    mVHACD = VHACD::CreateVHACD();
+
+    mVHACDParameters.m_callback = &mVHACDCallback;
+    mVHACDParameters.m_logger = &mVHACDLogger;
+
+    mDecompStages[0].mName = "Analyze";
+    mDecompStages[0].mDescription = nullptr;
+
+    LLCDParam param;
+    param.mName = "Fill Mode";
+    param.mDescription = nullptr;
+    param.mType = LLCDParam::LLCD_ENUM;
+    param.mDetails.mEnumValues.mNumEnums = 3;
+
+    static LLCDParam::LLCDEnumItem fill_enums[3];
+    fill_enums[(size_t)VHACD::FillMode::FLOOD_FILL].mName = "Flood";
+    fill_enums[(size_t)VHACD::FillMode::FLOOD_FILL].mValue = (int)VHACD::FillMode::FLOOD_FILL;
+    fill_enums[(size_t)VHACD::FillMode::SURFACE_ONLY].mName = "Surface Only";
+    fill_enums[(size_t)VHACD::FillMode::SURFACE_ONLY].mValue = (int)VHACD::FillMode::SURFACE_ONLY;
+    fill_enums[(size_t)VHACD::FillMode::RAYCAST_FILL].mName = "Raycast";
+    fill_enums[(size_t)VHACD::FillMode::RAYCAST_FILL].mValue = (int)VHACD::FillMode::RAYCAST_FILL;
+
+    param.mDetails.mEnumValues.mEnumsArray = fill_enums;
+    param.mDefault.mIntOrEnumValue = (int)VHACD::FillMode::FLOOD_FILL;
+    param.mStage = 0;
+    param.mReserved = -1;
+    mDecompParams.push_back(param);
+
+    enum EVoxelQualityLevels
+    {
+        E_LOW_QUALITY = 0,
+        E_NORMAL_QUALITY,
+        E_HIGH_QUALITY,
+        E_VERY_HIGH_QUALITY,
+        E_ULTRA_QUALITY,
+        E_MAX_QUALITY,
+        E_NUM_QUALITY_LEVELS
+    };
+
+    param.mName = "Voxel Resolution";
+    param.mDescription = nullptr;
+    param.mType = LLCDParam::LLCD_ENUM;
+    param.mDetails.mEnumValues.mNumEnums = E_NUM_QUALITY_LEVELS;
+
+    static LLCDParam::LLCDEnumItem voxel_quality_enums[E_NUM_QUALITY_LEVELS];
+    voxel_quality_enums[E_LOW_QUALITY].mName = "Low";
+    voxel_quality_enums[E_LOW_QUALITY].mValue = 200000;
+    voxel_quality_enums[E_NORMAL_QUALITY].mName = "Normal";
+    voxel_quality_enums[E_NORMAL_QUALITY].mValue = 400000;
+    voxel_quality_enums[E_HIGH_QUALITY].mName = "High";
+    voxel_quality_enums[E_HIGH_QUALITY].mValue = 800000;
+    voxel_quality_enums[E_VERY_HIGH_QUALITY].mName = "Very High";
+    voxel_quality_enums[E_VERY_HIGH_QUALITY].mValue = 1200000;
+    voxel_quality_enums[E_ULTRA_QUALITY].mName = "Ultra";
+    voxel_quality_enums[E_ULTRA_QUALITY].mValue = 1600000;
+    voxel_quality_enums[E_MAX_QUALITY].mName = "Maximum";
+    voxel_quality_enums[E_MAX_QUALITY].mValue = 2000000;
```

Comment:
You could tidy this up, make it setup entirely constexpr, and get rid of `EVoxelQualityLevels`, as the size and the index values will automatic.
```suggestion
    static constexpr LLCDParam::LLCDEnumItem voxel_quality_enums[] = {
        {"Low", 200000},
        {"Normal", 400000},
        {"High", 800000},
        {"Very High", 1200000},
        {"Ultra", 1600000},
        {"Maximum", 2000000}
    };
```